src/librustc_infer/traits/util.rs

```diff
@@ -117,35 +117,60 @@ pub fn elaborate_predicates<'tcx>(
 ) -> Elaborator<'tcx> {
     let mut visited = PredicateSet::new(tcx);
     predicates.retain(|pred| visited.insert(pred));
-    Elaborator { stack: predicates, visited }
+    let obligations: Vec<_> =
+        predicates.into_iter().map(|predicate| predicate_obligation(predicate, None)).collect();
+    elaborate_obligations(tcx, obligations)
+}
+
+pub fn elaborate_obligations<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    mut obligations: Vec<PredicateObligation<'tcx>>,
+) -> Elaborator<'tcx> {
+    let mut visited = PredicateSet::new(tcx);
+    obligations.retain(|obligation| visited.insert(&obligation.predicate));
+    Elaborator { stack: obligations, visited }
+}
+
+fn predicate_obligation<'tcx>(
+    predicate: ty::Predicate<'tcx>,
+    span: Option<Span>,
+) -> PredicateObligation<'tcx> {
+    let mut cause = ObligationCause::dummy();
+    if let Some(span) = span {
+        cause.span = span;
+    }
+    Obligation { cause, param_env: ty::ParamEnv::empty(), recursion_depth: 0, predicate }
 }
```

Comment:
Might be easier, instead of all of this, to have `Elaborator` produce `(ty::Predicate, Span)` pairs, then it's basically just like `tcx.predicates_of`.

Or `Option<Span>` I guess.

Reply:
I thought about those possibilities, but it felt that just dealing with `Obligation`s everywhere will make more sense in the future. If we can move to a point where we only deal with bare `Predicate` only when evaluating a single one, the we can keep as much diagnostic info as we (really) need.

Reply:
I suppose, but you're introducing dummy `Obligation`s that could end up getting mixed with true `Obligation`s.

This may make more sense if `elaborate_predicates` took in `Obligation<Vec<Predicate>>` or something.

But I suppose I can defer to @nikomatsakis on this.